jctools-core/src/main/java/org/jctools/queues/MpscBlockingConsumerArrayQueue.java

```diff
@@ -364,17 +364,32 @@ public E take() throws InterruptedException
             long pIndex = lvProducerIndex();
             if (cIndex == pIndex && casProducerIndex(pIndex, pIndex + 1))
             {
-                // producers only try a wakeup when both the index and the blocked thread are visible
-                soBlocked(Thread.currentThread());
-                do
+                boolean unblocked = false;
+                try
                 {
-                    LockSupport.park();
-                    if (Thread.currentThread().isInterrupted())
+                    // producers only try a wakeup when both the index and the blocked thread are visible
+                    soBlocked(Thread.currentThread());
+                    do
                     {
-                        throw new InterruptedException();
+                        LockSupport.park();
+                        if (Thread.interrupted())
+                        {
+                            throw new InterruptedException();
+                        }
+                    }
+                    while (lvBlocked() != null);
+                    unblocked = true;
+                }
+                finally
+                {
+                    if (!unblocked) {
```

Comment:
I think the diff would be cleaner/clearer if the boolean is removed and instead we incorporate the logic into the `Thread.interrupted` check. WDYT?

Reply:
Thanks @nitsanw I agree and actually that was my first pass, but the rationale for this approach is:
- It's more obviously robust, in that it's clear the queue is always left in a non-corrupt state without having to inspect and reason about what other obscure `Error`s may or may not be possible
- It nicely generalizes to the other `poll(long,TimeUnit)` PR impl, where there's a second "normal" abort condition, and seemed sensible for the logic to be consistent between the two

But I'm of course happy to go with whatever you prefer.

PS big fan of your work!